### PR TITLE
Explore/Logs: Escaping of incorrectly escaped log lines

### LINKED ISSUE #27964: Logs: Provide option to toggle \n and \t characters in log messages to literal newlines and tabs
Per https://github.com/grafana/loki/issues/2562#issuecomment-700237334, I am moving this issue from Loki repo to Grafana repo:

**Is your feature request related to a problem? Please describe.**
Many logs (e..g., from java or SQL) will include log messages that embed a lot of `\n` or `\r\n` or `\t` characters. These formatted log messages are pretty hard to read when viewed in Loki. They would be much easier to read if we could toggle into a view where these formatting characters are rendered as actual newlines & tabs.

Example screenshot:
image

**Describe the solution you'd like**
The Loki UI depends somewhat on a "one log entry per line" view. When you click on a log line, you get a pop-out with the log labels and parsed fields.

My idea is to provide a tab or toggle in the Loki log UI that allows the user to toggle into a view where the log message is displayed with literal newlines and tabs substituted for the `\n` and `\t` characters.

**Describe alternatives you've considered**
I have read through https://github.com/grafana/loki/issues/74, which discusses a related-but-different problem: i.e., that is basically the inverse problem: how to parse logging in such a way that a multi-line entry (like a stack trace) is treated as a single log entry.

I do not see any other existing GH issues that treat this exact same feature request.

**Additional context**
Here is an example of one log line pulled from Loki, which shows the kind of formatting that is typical in some logging:
[one-log-line_prettified.JSON.txt](https://github.com/grafana/grafana/files/5311958/one-log-line_prettified.JSON.txt)

and here is how that **message** looks like after substituting literal newlines for every "\n":
[one-log-line-MESSAGE-with-newlines.txt](https://github.com/grafana/grafana/files/5311964/one-log-line-MESSAGE-with-newlines.txt)

### PR DESCRIPTION
**Pull request by ivanahuckova**
Over the time, we've had a decent amount of reported issues related to escaping of newlines. Wast majority of those was due to the fact, that users had incorrectly escaped log lines. In that case, there is not much we can do at client side - at Grafana, as we don't know which characters are actual newlines and which have just the `\n` pattern.

@ethirolle [came up with and documented a good idea](https://github.com/grafana/grafana/issues/27964)  to give users a chance to replace their incorrectly escaped newlines and tabs with simply replacing `\t`,  `\n`,  `\r`, with actual newline or tab. 

**❗️What is very important** - we should communicate to users that the lines aren't correctly escaped and that the first step should be fixing escaping. However, for some logs it is very hard/almost impossible to correctly escape them. And for those cases, we would provide them with a feature that would replace incorrectly escapes newlines with actual newlines. Of course, it might always produce incorrect results, as we don't know which characters are actual newlines and which have just the `\n` pattern, but I think that if we communicate it clearly, users will understand. 

In this PR, I am showing the Escaping button only when we detect `\t`,  `\n`,  `\r` in any of the log lines. It has also experimental feature tooltip. We are also escaping and re-rendering only those lines, that are incorrectly escaped. 

https://user-images.githubusercontent.com/30407135/108508913-d3eee900-72bc-11eb-9c4b-1812a8e5aa87.mov

image

When no incorrectly escaped lines - no button:
image

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/grafana/issues/27964

**Special notes for your reviewer**:

This is the POC/my hack day project.

### CODE REVIEW

File: public/app/features/explore/Logs.tsx [hunk 1]
```diff
@@ -123,6 +135,17 @@ export class UnthemedLogs extends PureComponent<Props, State> {
     this.cancelFlippingTimer = setTimeout(() => this.setState({ isFlipping: false }), 1000);
   };
 
+  onEscapeNewlines = () => {
+    this.setState({ isEscaping: true });
```
Comment (aocenas) [inline public/app/features/explore/Logs.tsx:139]: Not sure I get why this is here. Does it take such a long time to rerender when escaping that we want to have essentially a loading state? If so does that mean it blocks all other actions?

**Comment by aocenas** [comment]
I like this, seems like a useful thing and I like that it's visible only with needed with good copy explaining why it's there.

@davkal @jessover9000 what do you think?

File: public/app/features/explore/Logs.tsx (hunk 1 above)
Comment (ivanahuckova) [reply public/app/features/explore/Logs.tsx:139]: Not really. It escapes the lines right away. I just liked that it visually showed you that something is happening. The same way we are doing it in fliping order. But the loading logic could be removed.

Comment (ivanahuckova) [reply public/app/features/explore/Logs.tsx]: But yeah, you are right. Too much code with a little/no effect. I have removed it.

File: packages/grafana-ui/src/components/Logs/LogRow.tsx [hunk 2]
```diff
@@ -42,14 +43,15 @@ interface Props extends Themeable {
   timeZone: TimeZone;
   allowDetails?: boolean;
   logsSortOrder?: LogsSortOrder | null;
+  escapedNewlines?: boolean;
```
Comment (Elfo404) [inline packages/grafana-ui/src/components/Logs/LogRow.tsx:46]: what about `forceEscape` instead?
or something else in an "active" form, like `showEscaped`, would it make it clearer?

File: public/app/features/explore/Logs.tsx [hunk 3]
```diff
@@ -346,6 +367,27 @@ export class UnthemedLogs extends PureComponent<Props, State> {
           />
         )}
 
+        {hasUnescapedNewlines && (
+          <MetaInfoText
+            metaItems={[
+              {
+                label: 'Your logs might have incorrectly escaped newlines',
```
Comment (Elfo404) [inline public/app/features/explore/Logs.tsx:374]: maybe, as it's also about tabs:

```suggestion
                label: 'Your logs might have incorrectly escaped content',
```

File: packages/grafana-data/src/types/logs.ts [hunk 4]
```diff
@@ -62,6 +62,7 @@ export interface LogRowModel {
   // Actual log line
   entry: string;
   hasAnsi: boolean;
+  hasUnescapedNewlines: boolean;
```
Comment (Elfo404) [inline packages/grafana-data/src/types/logs.ts:65]: what about `hasUnescapedContent` (we are also applying this to `\t` )

**Review by Elfo404 — DISMISSED** [review]
I played with it a bit and seems to work like a charm! 🚀 🔥 

I only have some doubts regarding some phrasing/naming for which I added a few comments.

I also couldn't see any major perf issue, did you run some tests as well?

EDIT: Also, I'm not sure if the button can be instead a switch placed beside all the other options, it would make things a bit more consistent but it's starting to get crowded in there, so I'd wait for more UI/UX feedback.

**Comment by Elfo404** [comment]
I found an issue when using this log line: `"Lorem \u001B[31mipsum \\n et\u001B[0m dolor"`:

https://user-images.githubusercontent.com/1170767/109817038-56669980-7c29-11eb-966f-bc26ac0fbc36.mov

which also happens if the unescaped `\n` is outside of the ANSI codes: 

Screenshot 2021-03-03 at 14 06 05

but works fine when there's no ANSI code in the string:

https://user-images.githubusercontent.com/1170767/109817483-e3115780-7c29-11eb-8a89-5827b7174310.mov


Also, which is a minor thing probably, as It's a very super-edge case, if I change the time range to only show a subset of logs that don't need escaping, the button will still be there. the only way to make it go away is to refresh the page. I don't know if this was intended, but it might be confusing:


https://user-images.githubusercontent.com/1170767/109818197-a2660e00-7c2a-11eb-83e6-e2d5514f79b1.mov

**Review by Elfo404 — APPROVED** [review]

**Comment by sourcehawk** [comment]
Is this option only available in the exporer feature? I cannot find any documentation on this anywhere and I would like to be able to view my logs without copy pasting them into a text file...

**Pull request merged**